Tests reliably failing on Python 3.7

<img width="819" alt="image" src="https://user-images.githubusercontent.com/9599/188504608-5be0f941-dca2-4f42-83bc-3d04bbad221d.png">

https://github.com/simonw/datasette/runs/8194907739?check_suite_focus=true

I thought this might be an intermittent failure but attempts to re-run the tests have not made it pass.

End of that trace is:

```
/home/runner/work/datasette/datasette/datasette/app.py:234: in __init__
    self._refresh_schemas_lock = asyncio.Lock()
/opt/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/asyncio/locks.py:161: in __init__
    self._loop = events.get_event_loop()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <asyncio.unix_events._UnixDefaultEventLoopPolicy object at 0x7fb1fc799fd0>

    def get_event_loop(self):
        """Get the event loop for the current context.
    
        Returns an instance of EventLoop or raises an exception.
        """
        if (self._local._loop is None and
                not self._local._set_called and
                isinstance(threading.current_thread(), threading._MainThread)):
            self.set_event_loop(self.new_event_loop())
    
        if self._local._loop is None:
            raise RuntimeError('There is no current event loop in thread %r.'
>                              % threading.current_thread().name)
E           RuntimeError: There is no current event loop in thread 'MainThread'.
```

--- a/datasette/app.py
+++ b/datasette/app.py
@@ -234,6 +234,8 @@ class Datasette:
         try:
             self._refresh_schemas_lock = asyncio.Lock()
         except RuntimeError as rex:
+            # Workaround for intermittent test failure, see:
+            # https://github.com/simonw/datasette/issues/1802
             if "There is no current event loop in thread" in str(rex):
                 loop = asyncio.new_event_loop()
                 asyncio.set_event_loop(loop)

--- a/datasette/app.py
+++ b/datasette/app.py
@@ -231,7 +231,13 @@ class Datasette:
         self.inspect_data = inspect_data
         self.immutables = set(immutables or [])
         self.databases = collections.OrderedDict()
-        self._refresh_schemas_lock = asyncio.Lock()
+        try:
+            self._refresh_schemas_lock = asyncio.Lock()
+        except RuntimeError as rex:
+            if "There is no current event loop in thread" in str(rex):
+                loop = asyncio.new_event_loop()
+                asyncio.set_event_loop(loop)
+                self._refresh_schemas_lock = asyncio.Lock()
         self.crossdb = crossdb
         self.nolock = nolock
         if memory or crossdb or not self.files:

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -22,13 +22,6 @@ from unittest import mock
 import urllib
 
 
-@pytest.fixture
-def ensure_eventloop():
-    # Workaround for "Event loop is closed" error
-    if asyncio.get_event_loop().is_closed():
-        asyncio.set_event_loop(asyncio.new_event_loop())
-
-
 def test_inspect_cli(app_client):
     runner = CliRunner()
     result = runner.invoke(cli, ["inspect", "fixtures.db"])
@@ -72,7 +65,7 @@ def test_serve_with_inspect_file_prepopulates_table_counts_cache():
     ),
 )
 def test_spatialite_error_if_attempt_to_open_spatialite(
-    ensure_eventloop, spatialite_paths, should_suggest_load_extension
+    spatialite_paths, should_suggest_load_extension
 ):
     with mock.patch("datasette.utils.SPATIALITE_PATHS", spatialite_paths):
         runner = CliRunner()
@@ -199,14 +192,14 @@ def test_version():
 
 
 @pytest.mark.parametrize("invalid_port", ["-1", "0.5", "dog", "65536"])
-def test_serve_invalid_ports(ensure_eventloop, invalid_port):
+def test_serve_invalid_ports(invalid_port):
     runner = CliRunner(mix_stderr=False)
     result = runner.invoke(cli, ["--port", invalid_port])
     assert result.exit_code == 2
     assert "Invalid value for '-p'" in result.stderr
 
 
-def test_setting(ensure_eventloop):
+def test_setting():
     runner = CliRunner()
     result = runner.invoke(
         cli, ["--setting", "default_page_size", "5", "--get", "/-/settings.json"]
@@ -215,14 +208,14 @@ def test_setting(ensure_eventloop):
     assert json.loads(result.output)["default_page_size"] == 5
 
 
-def test_setting_type_validation(ensure_eventloop):
+def test_setting_type_validation():
     runner = CliRunner(mix_stderr=False)
     result = runner.invoke(cli, ["--setting", "default_page_size", "dog"])
     assert result.exit_code == 2
     assert '"default_page_size" should be an integer' in result.stderr
 
 
-def test_config_deprecated(ensure_eventloop):
+def test_config_deprecated():
     # The --config option should show a deprecation message
     runner = CliRunner(mix_stderr=False)
     result = runner.invoke(
@@ -233,14 +226,14 @@ def test_config_deprecated(ensure_eventloop):
     assert "will be deprecated in" in result.stderr
 
 
-def test_sql_errors_logged_to_stderr(ensure_eventloop):
+def test_sql_errors_logged_to_stderr():
     runner = CliRunner(mix_stderr=False)
     result = runner.invoke(cli, ["--get", "/_memory.json?sql=select+blah"])
     assert result.exit_code == 1
     assert "sql = 'select blah', params = {}: no such column: blah\n" in result.stderr
 
 
-def test_serve_create(ensure_eventloop, tmpdir):
+def test_serve_create(tmpdir):
     runner = CliRunner()
     db_path = tmpdir / "does_not_exist_yet.db"
     assert not db_path.exists()
@@ -258,7 +251,7 @@ def test_serve_create(ensure_eventloop, tmpdir):
     assert db_path.exists()
 
 
-def test_serve_duplicate_database_names(ensure_eventloop, tmpdir):
+def test_serve_duplicate_database_names(tmpdir):
     "'datasette db.db nested/db.db' should attach two databases, /db and /db_2"
     runner = CliRunner()
     db_1_path = str(tmpdir / "db.db")
@@ -273,7 +266,7 @@ def test_serve_duplicate_database_names(ensure_eventloop, tmpdir):
     assert {db["name"] for db in databases} == {"db", "db_2"}
 
 
-def test_serve_deduplicate_same_database_path(ensure_eventloop, tmpdir):
+def test_serve_deduplicate_same_database_path(tmpdir):
     "'datasette db.db db.db' should only attach one database, /db"
     runner = CliRunner()
     db_path = str(tmpdir / "db.db")
@@ -287,7 +280,7 @@ def test_serve_deduplicate_same_database_path(ensure_eventloop, tmpdir):
 @pytest.mark.parametrize(
     "filename", ["test-database (1).sqlite", "database (1).sqlite"]
 )
-def test_weird_database_names(ensure_eventloop, tmpdir, filename):
+def test_weird_database_names(tmpdir, filename):
     # https://github.com/simonw/datasette/issues/1181
     runner = CliRunner()
     db_path = str(tmpdir / filename)

